File: Binary_Search_Tree/node.js
function Node(value) {
  this.value = value;
  this.left = null;
  this.right = null;
}

Node.prototype.search = function (value) {
  if (this.value == value) {
    console.log("Found: ", value);
    return this;
  }

  if (value < this.value && this.left != null) {
    return this.left.search(value);
  }

  if (value > this.value && this.right != null) {
    return this.right.search(value);
  }

  console.log("Value not in the Tree");
  return null;
};

Node.prototype.visit = function () {
  if (this.left != null) {
    this.left.visit();
  }
  console.log(this.value);
  if (this.right != null) {
    this.right.visit();
  }
};

Node.prototype.addNode = function (n) {
  if (n.value < this.value && this.left == null) {
    return (this.left = n);
  }
  if (n.value < this.value && this.left != null) {
    return this.left.addNode(n);
  }

  if (n.value > this.value && this.right == null) {
    return (this.right = n);
  }
  if (n.value > this.value && this.right != null) {
    return this.right.addNode(n);
  }
};

File: Binary_Search_Tree/python/node.py
from __future__ import annotations
from typing import Optional


class Node:

    def __init__(self, value) -> None:
        self.value = value
        self.left = None
        self.right = None

    def add_node(self, node: Node) -> None:
        if node.value < self.value and self.left is None:
            self.left = node
            return

        if node.value >= self.value and self.right is None:
            self.right = node
            return

        if node.value < self.value:
            self.left.add_node(node)
            return

        if node.value >= self.value:
            self.right.add_node(node)
            return

    def traverse(self) -> None:
        if not self.left is None:
            self.left.traverse()
        print(self.value)
        if not self.right is None:
            self.right.traverse()

    def search(self, value: int) -> Optional[Node]:
        if self.value == value:
            return self

        if value < self.value and self.left is not None:
            return self.left.search(value)

        if value > self.value and self.right is not None:
            return self.right.search(value)

        return None
